core/src/main/scala/codesearch/core/lexer/Tokenizer.scala

```diff
@@ -0,0 +1,30 @@
+package codesearch.core.lexer
+
+import fastparse._
+import NoWhitespace._
+import codesearch.core.lexer.tokens._
+
+sealed trait TokenizerMethods {
+  protected def leftForOtherSymbols[_: P] = P(CharIn("(", "[", "[^")).!
+
+  protected def rightForOtherSymbols[_: P] = P(CharIn(")", "]")).!
+
+  protected def specialSymbols[_: P] = P(CharIn(" ", "$", "%", "^", "&", "*", "+", "?", "!", "")).!
+
+  protected def parserAnyStringInOtherSymbols[_: P] = P(!rightForOtherSymbols ~ AnyChar).rep.!
+
+  protected def parserOtherSymbols[_: P] = P(leftForOtherSymbols ~ parserAnyStringInOtherSymbols ~ rightForOtherSymbols).rep(1).!.map(Other(_))
+
+  protected def parserSpecialSymbol[_: P] = P(specialSymbols.map(specialSymbolInString => SpecialSymbol(specialSymbolInString.charAt(0))))
+
+  protected def parserAnyStringToSpecialSymbol[_: P] = P((!specialSymbols ~ AnyChar).rep(1).!.map(Content(_)))
+
+  protected def parseStringWithSpecialSymbols[_: P] = P(parserOtherSymbols | parserAnyStringToSpecialSymbol | parserSpecialSymbol).rep
+}
+
+object Tokenizer extends TokenizerMethods {
+  def parseStringWithSpecialSymbols(query: String): Seq[Token] = {
+    val Parsed.Success(tokens, _) = parse(query, parseStringWithSpecialSymbols(_))
```

Comment:
Are you sure that parsing will success?

Reply:
Yes, in any case, the result will be `Seq [Token]`. But I still will check!

Reply:
Even if regex has an invalid format?

Reply:
If the regex is malformed, we'd like to notify the user about that. So `parseStringWithSpecialSymbols` should have a way to signal failure, and then the failure should be propagated to the frontend.

I would suggest making this a separate task. For now we can build the parser in such a way that it would never fail – possibly by introducing a new token type, e.g. `Malformed`. Though first I'd like to see examples where we can already detect malformed regexes.

If we can't detect malformed regexes and the parser is built in such a way that it always succeeds, I think it's fine to throw error 500 if this assumption is violated.